core/vm/validate.go

```diff
@@ -0,0 +1,214 @@
+func validateCode(code []byte, section int, metadata []*FunctionMetadata, jt *JumpTable) error {
+	var (
+		i = 0
+		// Tracks the number of actual instructions in the code (e.g.
+		// non-immediate values). This is used at the end to determine
+		// if each instruction is reachable.
+		count    = 0
+		op       OpCode
+		analysis bitvec
+	)
+	// This loop visits every single instruction and verifies:
+	// * if the instruction is valid for the given jump table.
+	// * if the instruction has an immediate value, it is not truncated.
+	// * if performing a relative jump, all jump destinations are valid.
+	// * if changing code sections, the new code section index is valid and
+	//   will not cause a stack overflow.
+	for i < len(code) {
+		count++
+		op = OpCode(code[i])
+		if jt[op].undefined {
+			return fmt.Errorf("use of undefined opcode %s", op)
+		}
+		switch {
+		case op >= PUSH1 && op <= PUSH32:
+			size := int(op - PUSH0)
+			if len(code) <= i+size {
+				return fmt.Errorf("truncated operand")
+			}
+			i += size
+		case op == RJUMP || op == RJUMPI:
+			if len(code) <= i+2 {
+				return fmt.Errorf("truncated rjump* operand")
+			}
+			if err := checkDest(code, &analysis, i+1, i+3, len(code)); err != nil {
+				return err
+			}
+			i += 2
+		case op == RJUMPV:
+			if len(code) <= i+1 {
+				return fmt.Errorf("truncated jump table operand")
+			}
+			count := int(code[i+1])
+			if count == 0 {
+				return fmt.Errorf("rjumpv branch count must not be 0")
+			}
+			if len(code) <= i+count {
+				return fmt.Errorf("truncated jump table operand")
+			}
+			for j := 0; j < count; j++ {
+				if err := checkDest(code, &analysis, i+2+j*2, i+2*count+2, len(code)); err != nil {
+					return err
+				}
+			}
+			i += 1 + 2*count
+		case op == CALLF:
+			if i+2 >= len(code) {
+				return fmt.Errorf("truncated operand")
+			}
+			arg, _ := parseUint16(code[i+1:])
+			if arg >= len(metadata) {
+				return fmt.Errorf("code section out-of-bounds (want: %d, have: %d)", arg, len(metadata))
+			}
+			i += 2
+		}
+		i += 1
+	}
+	// Code sections may not "fall through" and require proper termination.
+	// Therefore, the last instruction must be considered terminal.
+	if !jt[op].terminal {
+		return fmt.Errorf("code section ends with non-terminal instruction")
+	}
+	if paths, err := validateControlFlow(code, section, metadata, jt); err != nil {
+		return err
+	} else if paths != count {
+		return fmt.Errorf("unreachable code")
+	}
+	return nil
+}
+
+// checkDest parses a relative offset at code[0:2] and checks if it is a valid jump destination.
+func checkDest(code []byte, analysis *bitvec, imm, from, length int) error {
+	if len(code) < imm+2 {
+		return fmt.Errorf("truncated operand")
+	}
+	if analysis != nil && *analysis == nil {
+		*analysis = eofCodeBitmap(code)
+	}
+	dest := from + parseInt16(code[imm:])
+	if dest < 0 || dest >= length {
+		return fmt.Errorf("relative offset out-of-bounds: %d", dest)
```

Comment:
```suggestion
		return fmt.Errorf("destination out-of-bounds: %d", dest)
```
The `dest` is not the relative offset, you have already made it absolute by adding the arg to `from`